### ISSUE TITLE
Disambiguate error messages from EqualConstraint

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#1808]
Although we try to avoid it, EqualConstraint occasionally fails with an error message like

```
Expected: some_value
 But was: some_value <== Same value in both messages
```

Issue #1778 illustrates such a case, producing the message

```
Expected: 0
 But was: 0
```

It is proposed that we append the type in such a case.,.

```
Expected: 0 (System.Int32)
 But was: 0 (System.IntPtr)
```

Note this has to be done somewhere in the code where **both** the expected and the actual values are available and after the string representations have already been determined. Finding the right spot may be slightly tricky.

Notes:
1. We should only do this if the types are actually different, otherwise, do not bother.
2. If a tolerance or other constraint modifier is in effect, we should investigate whether to enable this feature. The best course may be different for each modifier.

**Comment by oznetmaster** [comment]
If this can be done, it would be great!

**Comment by jplebre** [comment]
What would happen if namespaces are incredibly long?
(Mysolution.myproject.tests.helpers.factory.stub.consumeraddress)
Or something silly like that.

Could it be worth to just have the type?

**Comment by CharliePoole** [comment]
One could just use the name in most cases, I think.

**Comment by oznetmaster** [comment]
Just use as much of the name as needed to make it unique (right to left).

**Comment by JustinRChou** [comment]
Tentatively I wrote the following to get the difference type names. Still looking into how to get it into the Write function naturally.
```
        private void ResolveTypeNameDifference(object expected, object actual, out string expectedType, out string actualType) {
            string[] expectedOriginalType = expected.GetType().ToString().Split('.'); //NUnit.Framework.Constraints.Outer.Middle.Inner.Class1
            string[] actualOriginalType = actual.GetType().ToString().Split('.'); //NUnit.Framework.Constraints.Outer.Hey.Inner.Class1
            int actualStart = 0, expectStart = 0;
            for (int expectLen = expectedOriginalType.Length - 1,actualLen = actualOriginalType.Length - 1;
                expectLen >= 0 && actualLen >= 0;
                expectLen--, actualLen--) {
                if (expectedOriginalType[expectLen] != actualOriginalType[actualLen]) {
                    actualStart = actualLen;
                    expectStart = expectLen;
                    break;
                }
            }
            expectedType = String.Join(".", expectedOriginalType, expectStart, expectedOriginalType.Length- expectStart); //Middle.Inner.Class1
            actualType = String.Join(".", actualOriginalType, actualStart, actualOriginalType.Length- actualStart); //Hey.Inner.Class1
        }
```

**Comment by JustinRChou** [comment]
While I am in the middle of figuring out to create a pull request for this.

But my update did cause failures in other tests asserting on the exception message.

Would this be acceptable since it would mess up a lot of other users asserts?

**Comment by CharliePoole** [comment]
@JustinRChou Your changes should not break other tests, since we only want to show the type when there is ambiguity and I don't think there is ambiguity in any of our existing tests. Of course, if there is, then your change will impact them and that's OK, but only in that case.

Your code is a bit hard to follow out of context. You should create a feature branch to work on this and show where you will make the change. I suggest doing the simplest thing first and only worrying about truncating type names after the simple version is working and then only if it looks like we need it.

**Comment by JustinRChou** [comment]
Updated the PR with Fixes.

Since it was merged so I am closing this.
